### PR TITLE
Ethernet: Remove V1 code for stm32H7 and stm32H5 series

### PR DESCRIPTION
**Pull request by marwaiehm-st**
- Add stm32h7 ethernet compatible "st,stm32h7-ethernet", used also for stm32h5.
- Remove the V1 code for STM32H7 and STM32H5 series
- Replace "CONFIG_SOC_SERIES_STM32H7X || CONFIG_SOC_SERIES_STM32H5X"
by the compatble st_stm32h7_ethernet

### CODE REVIEW

File: drivers/ethernet/eth_stm32_hal.c [hunk 1]
```diff
@@ -1486,27 +1296,30 @@ static const struct eth_stm32_hal_dev_cfg eth0_config = {
...
+					.PhyAddress = PHY_ADDR,
+					.RxMode = ETH_RXINTERRUPT_MODE,
+					.ChecksumMode = IS_ENABLED(CONFIG_ETH_STM32_HW_CHECKSUM)
+								? ETH_CHECKSUM_BY_HARDWARE
+								: ETH_CHECKSUM_BY_SOFTWARE,
+#endif /* !CONFIG_ETH_STM32_HAL_API_V2 */
+					.MediaInterface = IS_ENABLED(CONFIG_ETH_STM32_HAL_MII)
+								  ? ETH_MEDIA_INTERFACE_MII
+								  : ETH_MEDIA_INTERFACE_RMII,
+			},
```
Comment (erwango) [inline drivers/ethernet/eth_stm32_hal.c:1322]: I'm not sure about the indentation shift in this would section.
If really required, please make it in a dedicated commit.

File: drivers/ethernet/eth_stm32_hal.c [hunk 2]
```diff
@@ -1706,27 +1519,27 @@ static int ptp_stm32_init(const struct device *port)
...
-		(clock_control_subsys_t)&eth_cfg->pclken,
+#if DT_HAS_COMPAT_STATUS_OKAY(st_stm32h7_ethernet)
+				     (clock_control_subsys_t)&eth_cfg->pclken,
 #else
-		(clock_control_subsys_t)&eth_cfg->pclken_ptp,
-#endif /* CONFIG_SOC_SERIES_STM32H7X || CONFIG_SOC_SERIES_STM32H5X */
-		&ptp_hclk_rate);
+				     (clock_control_subsys_t)&eth_cfg->pclken_ptp,
+#endif /* DT_HAS_COMPAT_STATUS_OKAY(st_stm32h7_ethernet) */
+				     &ptp_hclk_rate);
```
Comment (erwango) [inline drivers/ethernet/eth_stm32_hal.c:1542]: Please don't mix indentation changes with the clean up in the same commit.

**Review by erwango — CHANGES_REQUESTED** [review]
Thanks, a quite welcome clean up. Some points to verify, though.

File: drivers/ethernet/eth_stm32_hal.c (hunk 1 above)
Comment (marwaiehm-st) [reply drivers/ethernet/eth_stm32_hal.c:1322]: It's required in the compilation checks. I can remove it.
Screenshot from 2024-10-18 12-01-54

File: drivers/ethernet/eth_stm32_hal.c [hunk 3]
```diff
@@ -394,37 +383,19 @@ static int eth_tx(const struct device *dev, struct net_pkt *pkt)
...
-	defined(CONFIG_ETH_STM32_HAL_API_V2)
-
-	tx_config.Length = total_len;
 #if defined(CONFIG_ETH_STM32_HAL_API_V2)
+	tx_config.Length = total_len;
 	tx_config.pData = &ctx;
 	tx_config.TxBuffer = &dma_tx_buffer_header[ctx.first_tx_buffer_index].tx_buff;
-#else
-	tx_config.TxBuffer = &tx_buffer_def;
-#endif /* CONFIG_ETH_STM32_HAL_API_V2 */
```
[the remark this answers is not in the document]
Comment (marwaiehm-st) [reply drivers/ethernet/eth_stm32_hal.c:427]: > Are you sure these lines should be removed? In non API V2 case, they will be missing and this makes the remaining part of the function API V2 specific, which doesn't look right. Or it should be added line 489.

Yes, it should be removed because it is included in the:

```c
#if defined(CONFIG_SOC_SERIES_STM32H7X) || defined(CONFIG_SOC_SERIES_STM32H5X) || \
	defined(CONFIG_ETH_STM32_HAL_API_V2)
``` 
so the else is not handling the case of all non-API V2, but only the case of:

```c
#if defined(CONFIG_SOC_SERIES_STM32H7X) || defined(CONFIG_SOC_SERIES_STM32H5X)
``` 
Now the V1 API support has been removed for STM32H7 and STM32H5, so there is no need for this.

[the remark this answers is not in the document]
Comment (marwaiehm-st) [reply drivers/ethernet/eth_stm32_hal.c:427]: > Note also that the code under line 452 `#ifndef CONFIG_ETH_STM32_HAL_API_V2` should now be removed.

Yes, I will remove it.

File: drivers/ethernet/eth_stm32_hal.c (hunk 1 above)
Comment (erwango) [reply drivers/ethernet/eth_stm32_hal.c:1322]: Ok, these are just warnings which won't affect result warning, so not mandatory to fix.
In general case, if you chose to fix them, please do it in a dedicated commit.

**Review by erwango — DISMISSED** [review]

File: drivers/ethernet/eth_stm32_hal.c [hunk 4]
```diff
@@ -65,7 +65,7 @@ static const struct device *eth_stm32_phy_dev = DEVICE_PHY_BY_NAME(0);
 
 #endif
 
-#if defined(CONFIG_SOC_SERIES_STM32H7X) || defined(CONFIG_SOC_SERIES_STM32H5X)
+#if DT_HAS_COMPAT_STATUS_OKAY(st_stm32h7_ethernet)
```
Comment (mathieuchopstm) [inline drivers/ethernet/eth_stm32_hal.c:68]: For readability, I think the replacement of `#if SOC_H7 || SOC_H5` with `HAS_COMPAT(h7_eth)` should be split to a separate commit.

**Review by mathieuchopstm — CHANGES_REQUESTED** [review]
Small details but otherwise LGTM.

File: dts/bindings/ethernet/st,stm32-ethernet.yaml [hunk 5]
```diff
@@ -1,22 +1,8 @@
-# Copyright (c) 2020, Linaro Limited
+# Copyright (c) 2024, STMicroelectronics
```
[the remark this answers is not in the document]
Comment (marwaiehm-st) [reply dts/bindings/ethernet/st,stm32-ethernet.yaml:1]: So, i will keep the original and the new one

File: drivers/ethernet/eth_stm32_hal.c (hunk 4 above)
Comment (marwaiehm-st) [reply drivers/ethernet/eth_stm32_hal.c:68]: ok, i will put it in a separate commit

**Review by erwango — APPROVED** [review]

**Review by mathieuchopstm — APPROVED** [review]
